draft-ietf-ppm-dap.md

```diff
@@ -1214,64 +1269,174 @@ as follows:
+  `prepare_steps` in the Leader's `AggregationJob` (if this is a subsequent
+  round).
+* The Helper MUST advance to the same round as the Leader. That is, the Helper's
+ `AggregationJob.round` MUST be 0 if this is the first round of continuation or
+  MUST be equal the Leader's `AggregationJob.round` if this is a subsequent
+  round).
+
+If either condition does not hold, the Leader should abandon further processing
+of the aggregation job.
+
+Once the Helper's `AggregationJob` is validated, the Leader can continue the
+preparation of each report share in the aggregation job based on its locally
+computed prepare message and the `PrepareStep` from the Helper.
+
+If `PrepareStep` is of type `failed`, then the Leader acts based on the value of
+the `ReportShareError`:
+
+* If the error is `ReportShareError.report_too_early`, then the Leader MAY try
+  to re-send the report in a later `AggregationJob`.
+* For any other error, the Leader marks the report as failed, removes it from
+  the candidate report set and does not process it further.
+
+If the Helper's `PrepareStep` is of type is `finished` and the Leader's
+preparation of this report share is also finished, then the report share is
+aggregated and can now be collected (see {{collect-flow}}). If the Leader is not
+finished, then the report cannot be processed further and MUST be abandoned.
+
+If the Helper's `PrepareStep` is of type `continued`, then the Leader proceeds
+as follows.
+
+Let `leader_outbound` denote the Leader's prepare message and `helper_outbound`
+denote the Helper's. The Leader advances its aggregation job to the next round
+by computes the next state transition as follows:
 
 ~~~
-enum {
-  continued(0),
-  finished(1),
-  failed(2),
-  (255)
-} PrepareStepResult;
+inbound = VDAF.prep_shares_to_prep(agg_param, [leader_outbound, helper_outbound])
+out = VDAF.prep_next(prep_state, inbound)
+~~~
 
-struct {
-  ReportID report_id;
-  PrepareStepResult prepare_step_result;
-  select (PrepareStep.prepare_step_result) {
-    case continued: opaque prep_msg<0..2^32-1>; /* VDAF preparation message */
-    case finished:  Empty;
-    case failed:    ReportShareError;
-  };
-} PrepareStep;
+where `[leader_outbound, helper_outbound]` is a vector of two elements. If
+either of these operations fails, then the Leader marks the report as invalid.
+Otherwise, if the VDAF has reached the finished state, then `out` will be the
+Leader's output share, in which case the Leader finishes and stores its output
+share for further processing as described in {{collect-flow}}. Otherwise, `out`
+is the pair `(new_state, prep_msg)`, where `new_state` is the Leader's updated
+state and `prep_msg` is the next VDAF message (which will be `leader_outbound`
+in the next round of continuation). For the latter case, the Leader sets
+`prep_state` to `new_state`.
+
+The Leader SHOULD store each report share's current preparation state and the
```

Comment:
Shouldn't this storage happen after receiving the response from the helper?

Reply:
I think there's a few different points at which the leader could store different things. What we care about is that the leader checkpoint state so that it can recover from crashes, but the more I think about it, the more I think it's inappropriate for the spec to say anything about what an implementation stores or when. So maybe we shouldn't say anything about this at all.

Reply:
We have the same problem in the equivalent helper text (noted by Brandon [here](https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/395#discussion_r1067587961). These two passages are awkward, because there's a few different ways for implementations to checkpoint state such that they can implement round skew recovery, and I don't think it's appropriate for the spec to dictate implementation details (because maybe there's some nuance about a particular implementation's storage system that makes the prescribed strategy difficult or impossible). Maybe these SHOULDs about storing state should be deleted and instead we could have an implementation note elsewhere (Operational Considerations?) discussing the intent of the `round` field. and suggesting one possible solution.